src/NuGet.Core/NuGet.Build.Tasks/AssignReferencePropertiesTask.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using NuGet.Commands;
+using NuGet.Frameworks;
+
+namespace NuGet.Build.Tasks
+{
+    public class AssignReferencePropertiesTask : Task
+    {
+        /// <summary>
+        /// The current project's target framework.
+        /// </summary>
+        [Required]
+        public string CurrentProjectTargetFramework { get; set; }
+
+        /// <summary>
+        /// The project references for property lookup.
+        /// </summary>
+        public ITaskItem[] AnnotatedProjectReferences { get; set; }
+
+        /// <summary>
+        /// The project references with assigned properties.
+        /// </summary>
+        [Output]
+        public ITaskItem[] AssignedProjects { get; set; }
+
+        public override bool Execute()
+        {
+            var log = new MSBuildLogger(Log);
+
+            BuildTasksUtility.LogInputParam(log, nameof(CurrentProjectTargetFramework), CurrentProjectTargetFramework);
+
+            if (AnnotatedProjectReferences != null)
+            {
+                log.LogDebug($"(in) AnnotatedProjectReferences '{string.Join(";", AnnotatedProjectReferences.Select(p => p.ItemSpec))}'");
+            }
+            else
+            {
+                return !Log.HasLoggedErrors;
+            }
+
+            var frameworkToMatch = NuGetFramework.Parse(CurrentProjectTargetFramework);
+            if (frameworkToMatch.IsUnsupported)
+            {
+                log.LogError(string.Format(Strings.UnsupportedTargetFramework, CurrentProjectTargetFramework));
+                return false;
+            }
+
+            AssignedProjects = AnnotatedProjectReferences.Select(project => AssignPropertiesForSingleReference(project, frameworkToMatch)).ToArray();
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private ITaskItem AssignPropertiesForSingleReference(ITaskItem project, NuGetFramework currentProjectTargetFramework)
+        {
+            var itemWithProperties = new TaskItem(project);
+
+            var targetFramework = project.GetMetadata("TargetFrameworks");
```

Comment:
nit: This should be named `targetFrameworks`

Reply:
Fixed.

---
In reply to: [143530143](https://github.com/NuGet/NuGet.Client/pull/1737#discussion_r143530143) [](ancestors = 143530143)